src/com/facebook/buck/worker/WorkerProcess.java

```diff
@@ -45,8 +55,14 @@
   private final Path stdErr;
   private final AtomicInteger currentMessageID = new AtomicInteger();
   private boolean handshakePerformed = false;
+  private final ConcurrentHashMap<Integer, SettableFuture<Integer>> commandExitCodes =
+      new ConcurrentHashMap<>();
   @Nullable private WorkerProcessProtocol.CommandSender protocol;
   @Nullable private ProcessExecutor.LaunchedProcess launchedProcess;
+  private final Thread readerThread;
+  private final CountDownLatch readerShutdownSignal = new CountDownLatch(1);
+  private final Thread watchdogThread;
+  private final CountDownLatch watchdogShutdownSignal = new CountDownLatch(1);
```

Comment:
The variables readerShutdownSignal and watchdogShutdownSignal are just flags, should we make them volatile boolean? That way we will have 2 less objects.

Reply:
Not quite - the latches are `.await`-able. Realistically these objects probably aren't memory-intensive for buck - there's a `WorkerProcess` object per long-lived process. You would rarely have more than ncpus of these processes running -- though I guess it depends how many worker tools you use!

Reply:
Thanks you